### PR TITLE
AWS IAM: Support for AWS EKS ServiceAccount roles for CloudWatch and S3 image upload

### LINKED ISSUE #20473: AWS IAM: Support for AWS EKS ServiceAccount roles
**What happened**:

When running Grafana in AWS EKS with IAM role attached to service account of a pod, Grafana still uses IAM role of an underlying EC2 instance. When adding Cloudwath datasource, the following error informs that Grafana uses instance role:

```
t=2019-11-19T10:31:39+0000 lvl=eror msg="Metric request error" logger=context userId=1 orgId=1 uname=admin error="Failed to call cloudwatch:ListMetrics, AccessDenied: User: arn:aws:sts::$ACCOUNT_ID:assumed-role/$INSTANCE_ROLE_NAME/$INSTANCE_ID is not authorized to perform: cloudwatch:ListMetrics\n\tstatus code: 403, request id: a4aac23f-2194-422a-83de-2926833ce40f”
```

$ACCOUNT_ID, $INSTANCE_ROLE_NAME and $INSTANCE_ID are sanitized from the output.

**What you expected to happen**:

Grafana would use IAM role attached to a service account of a pod.

**How to reproduce it (as minimally and precisely as possible)**:

* Setup OIDC provider for EKS
* Setup Kubernetes service account and create IAM role, which is attached to a service account
* Create Kubernetes deployment and define above-mentioned service account for pod

Walkthrough for setup steps https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/

**Anything else we need to know?**:

We also ran python container with same service account and installed aws-cli and it used the correct IAM role (not the underlying instance role):

```
# aws sts get-caller-identity
{
    "UserId": "$USER_ID:botocore-session-1574167643",
    "Account": "$ACCOUNT_ID",
    "Arn": "arn:aws:sts::$ACCOUNT_ID:assumed-role/grafana_test_eks_log_role/botocore-session-1574167643"
}
```

$USER_ID and $ACCOUNT_ID are sanitized from output.

awscli version: `aws-cli/1.16.284 Python/3.6.9 Linux/4.14.146-119.123.amzn2.x86_64 botocore/1.13.20`

In addition, the above-mentioned Grafana container has the following environment variables injected by admission controller:

```
/usr/share/grafana $ cat /proc/1/environ | tr '\0' '\n' |grep AWS
AWS_ROLE_ARN=arn:aws:iam::$ACCOUNT_ID:role/grafana_test_eks_log_role
AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/eks.amazonaws.com/serviceaccount/token
```

To my knowledge, this indicates that IAM attachment to service account and pod is done successfully.

To test this issue, we ran Grafana with minimal configuration.

PR https://github.com/grafana/grafana/pull/19138 updated aws-sdk-go to version which supports IAM roles to service accounts

**Environment**:
- Grafana version: v6.5.0-beta1 (`grafana/grafana:6.5.0-beta1` Docker image)
- Data source type & version: Cloudwatch
- OS Grafana is installed on: Alpine Linux 3.10.3 (Docker container)
- User OS & Browser: Chrome Version 78.0.3904.97 (Official Build) (64-bit)
- Grafana plugins: -
- Others:
  - EKS Kubernetes version: 1.14
  - EKS Platform version: eks.1
  - EKS worker node Kubernetes version: v1.14.7-eks-1861c5
  - Worker node AMI: Official EKS worker node AMI: amazon-eks-node-1.14-v20190927 (ami-059c6874350e63ca9) - Description: EKS Kubernetes Worker AMI with AmazonLinux2 image, (k8s: 1.14.7, docker:18.06)

### PR DESCRIPTION
**Pull request by patstrom**
**What this PR does / why we need it**:

This PR replicates (most of) the behavior of `aws-sdk-go` when fetching a WebIdentityRole such as they are used in IAM roles for service account in EKS.

The reason for the "(most of)" qualifier is that this will only read the environment variables, so if e.g a `role_session_name` is configured in a configuration file somewhere it won't be picked up. I suspect this won't be much of a problem since all documentation and whatnot for EKS IAM roles for Service Accounts exclusively uses those environment variables.

As discussed in the issue the "proper" way to fix this would probably be to always use the default credentials chain unless a specific way is specified in which case *only* that way should be used.

I've tried this out in an EKS cluster with a standard setup. In other words I've created a deployment and a service account, linked them together, annotated the service account with `eks.amazonaws.com/role-arn` and added the necessary trust policies on the role and I can confirm it manages to find the credentials.

**Which issue(s) this PR fixes**:

Fix #20473

### CODE REVIEW

**Comment by patstrom** [comment]
I realized this change is probably needed in the s3uploader for images as well. Perhaps the AWS configuration should be centralized into a single pkg that both the cloudwatch executor, the image uploader and any potential future interactions can re-use?

**Review by aknuds1 — CHANGES_REQUESTED** [review]
I made a couple of suggestions for minor improvements. I'll see if I can test this in order to reproduce the issue.

**Comment by aknuds1** [comment]
Thanks for applying my suggestions, I'm just waiting to get access to an EKS cluster so I can test.

File: pkg/tsdb/cloudwatch/credentials.go [hunk 1]
```diff
@@ -105,6 +106,7 @@ func GetCredentials(dsInfo *DatasourceInfo) (*credentials.Credentials, error) {
 			}},
 			&credentials.SharedCredentialsProvider{Filename: "", Profile: dsInfo.Profile},
 			remoteCredProvider(sess),
+			webIdentityProvider(sess),
```
Comment (evq) [inline pkg/tsdb/cloudwatch/credentials.go]: May also want `webIdentityProvider(stsSess)` below  https://github.com/grafana/grafana/blob/a78387fdc2176fe7a114958bcac91e2d4291dbf3/pkg/tsdb/cloudwatch/credentials.go#L63

**Comment by aknuds1** [comment]
Handing this over to @ryantxu since he needs to look it over too.

**Comment by patstrom** [comment]
@evq you need to sign the cla.

@ryanxtu what is the status on this?

**Comment by evq** [comment]
@patstrom signed

**Comment by iptizer** [comment]
Is there anything I can do to speed this up a little bit? Hit me!

How I see it @aknuds1 currently has some changes requested. Can I help you testing or something?

**Comment by patstrom** [comment]
I've already addressed the requested changes. Perhaps @aknuds1 needs to mark them as resolved as well?

**Review by aknuds1 — APPROVED** [review]
LGTM, but waiting for @ryantxu's review.

**Comment by patstrom** [comment]
@ryantxu What is the status on this?

**Comment by dnascimento** [comment]
@ryantxu  When may you review this?

**Comment by chancez** [comment]
How does this work in combination with configuring the assume role ARN for the cloudwatch datasource? Can we use the EKS service account as the source profile to assume role with the AWS IAM role specified in the datasource? I'm trying to see if once this is fixed if I can use the SA IAM role to allow assuming into a different IAM role per cloudwatch datasource so I can use a single grafana with multiple AWS accounts.

File: pkg/tsdb/cloudwatch/credentials.go [hunk 2]
```diff
@@ -60,6 +61,7 @@ func GetCredentials(dsInfo *DatasourceInfo) (*credentials.Credentials, error) {
 				&credentials.EnvProvider{},
 				&credentials.SharedCredentialsProvider{Filename: "", Profile: dsInfo.Profile},
 				remoteCredProvider(stsSess),
```
Comment (dnascimento) [inline pkg/tsdb/cloudwatch/credentials.go:63]: @patstromm, could you set `webIdentityProvider` before `remoteCredProvider`? It's likely that the EC2 instances have a role attached to it, so this order will prioritise the EC2 instance role over the webIdentity. It's better to try first to use the WebIdentity

**Review by dnascimento — CHANGES_REQUESTED** [review]

**Comment by dnascimento** [comment]
> How does this work in combination with configuring the assume role ARN for the cloudwatch datasource? Can we use the EKS service account as the source profile to assume role with the AWS IAM role specified in the datasource? I'm trying to see if once this is fixed if I can use the SA IAM role to allow assuming into a different IAM role per cloudwatch datasource so I can use a single grafana with multiple AWS accounts.

The EKS service account role can assume a role as well if you define the arn, the code flow is:
https://github.com/grafana/grafana/blob/42290dea178c4f721bede132be9ad2f8ba8dba81/pkg/tsdb/cloudwatch/credentials.go#L48
https://github.com/grafana/grafana/blob/42290dea178c4f721bede132be9ad2f8ba8dba81/pkg/tsdb/cloudwatch/credentials.go#L64
https://github.com/grafana/grafana/blob/42290dea178c4f721bede132be9ad2f8ba8dba81/pkg/tsdb/cloudwatch/credentials.go#L76

@patstrom I tested the code in our systems, could you address the 2 issues above so that this can be reviewed and merged?

**Comment by chancez** [comment]
@dnascimento yep, I see now, looks like it should work. I'll build this locally and see if I can get it to work with multiple roles. Thanks.

**Comment by patstrom** [comment]
I currently have a [WIP](https://github.com/patstrom/grafana/commit/f5ad86fc8bf17bf70eb108cf2d3ef56f7263e5c4) of the "proper" way here, where I add a `AWS SDK Default` authentication provider that does exactly what you'd expect. 

I plan to open a separate MR when it's ready (hopefully later this week).

**Comment by ryantxu** [comment]
@patstrom -- the new approach looks great.  thanks for the effort

**Comment by aknuds1** [comment]
Synced your branch with Grafana master, to fix CI issues.

**Comment by aknuds1** [comment]
This took a while, but we finally got it merged - thank you for contributing to Grafana!

**Comment by patstrom** [comment]
Cheers! :tada:

You're not getting rid of me that easily though. I just opened a [new PR](https://github.com/grafana/grafana/pull/23439) with the "proper" way to solve this that also includes some other fixes to the Cloudwatch data source executor.

**Pull request merged**